File: skills/blasphemous-modding-helper/scripts/test_modding_api_acceptance.js
"use strict";

const fs = require("fs");
const path = require("path");
const support = require("./modding_api_test_support");

const { repositoryRoot, scriptDirectory } = support;

function fail(message) {
  throw new Error(message);
}

function assert(condition, message) {
  if (!condition) {
    fail(message);
  }
}

function runSuite(label, invoke) {
  process.stdout.write(`[RUN] ${label}\n`);
  const result = invoke();
  if (result.status !== 0) {
    fail(
      `${label} failed with exit code ${result.status}${
        result.timedOut ? " (timed out)" : ""
      }:\n${result.output}`,
    );
  }
  process.stdout.write(`[OK] ${label}\n`);
}

function parseArguments(argv) {
  const options = { requireClean: false };
  for (const argument of argv) {
    if (argument === "--require-clean") {
      options.requireClean = true;
    } else if (argument === "--help" || argument === "-h") {
      process.stdout.write(
        "Usage: test_modding_api_acceptance.js [--require-clean]\n" +
          "  --require-clean  Fail unless the repository worktree is clean.\n",
      );
      return null;
    } else {
      fail(`unknown option: ${argument}`);
    }
  }
  return options;
}

function runResolverParity(bash, powerShell, env) {
  const fixtureRoot = support.createFixtureDirectory("modding-api-parity-");
  try {
    const metadataFile = path.join(fixtureRoot, "latest.json");
    fs.writeFileSync(
      metadataFile,
      JSON.stringify({
        tag_name: "v1.0.0",
        draft: false,
        prerelease: false,
        resolved_ref: "v1.0.0",
        resolved_commit: "0123456789012345678901234567890123456789",
      }),
      "utf8",
    );
    const resolverShell = path.join(scriptDirectory, "resolve_modding_api.sh");
    const resolverPowerShell = path.join(scriptDirectory, "resolve_modding_api.ps1");
    const bashResult = support.invokeBash(
      bash,
      resolverShell,
      ["--selector", "latest", "--metadata-file", metadataFile],
      env,
    );
    const powerShellResult = support.invokePowerShell(
      powerShell,
      resolverPowerShell,
      ["-Selector", "latest", "-MetadataFile", metadataFile],
      env,
    );
    assert(
      bashResult.status === 0 && powerShellResult.status === 0,
      `resolver parity fixture failed:\nBash:\n${bashResult.output}\nPowerShell:\n${powerShellResult.output}`,
    );
    const bashValues = support.parseKeyValues(bashResult.output);
    const powerShellValues = support.parseKeyValues(powerShellResult.output);
    for (const key of [
      "MODDING_API_REPOSITORY",
      "MODDING_API_SELECTOR",
      "MODDING_API_SELECTOR_KIND",
      "MODDING_API_RESOLVED_REF",
      "MODDING_API_RESOLVED_TAG",
      "MODDING_API_RESOLVED_COMMIT",
      "MODDING_API_DOCS_URL",
      "MODDING_API_SOURCE_URL",
    ]) {
      assert(
        bashValues[key] === powerShellValues[key],
        `resolver output mismatch for ${key}: Bash=${bashValues[key]} PowerShell=${powerShellValues[key]}`,
      );
    }

    const invalidBash = support.invokeBash(
      bash,
      resolverShell,
      ["--selector", "main"],
      env,
    );
    const invalidPowerShell = support.invokePowerShell(
      powerShell,
      resolverPowerShell,
      ["-Selector", "main"],
      env,
    );
    assert(
      invalidBash.status === 2 && invalidPowerShell.status === 2,
      `invalid selector exit-code mismatch: Bash=${invalidBash.status} PowerShell=${invalidPowerShell.status}`,
    );
    for (const output of [invalidBash.output, invalidPowerShell.output]) {
      assert(output.includes("[ERROR REPORT]"), "invalid selector must print an error report");
      assert(output.includes("next_step:"), "invalid selector report must include next_step");
    }
  } finally {
    fs.rmSync(fixtureRoot, { recursive: true, force: true });
  }
}

function runGit(workingDirectory, args) {
  const result = support.runCommand("git", ["-C", workingDirectory].concat(args));
  assert(result.status === 0, `Git fixture command failed: ${result.output}`);
  return result.output.trim();
}

function normalizedValue(values, key, replacements) {
  let value = values[key];
  assert(value !== undefined, `missing output field: ${key}`);
  for (const [source, replacement] of replacements) {
    value = value.split(source).join(replacement);
  }
  return value;
}

function compareSurfaceValues(label, bashOutput, powerShellOutput, keys, replacements) {
  const bashValues = support.parseKeyValues(bashOutput);
  const powerShellValues = support.parseKeyValues(powerShellOutput);
  for (const key of keys) {
    const bashValue = normalizedValue(bashValues, key, replacements);
    const powerShellValue = normalizedValue(powerShellValues, key, replacements);
    assert(
      bashValue === powerShellValue,
      `${label} output mismatch for ${key}: Bash=${bashValue} PowerShell=${powerShellValue}`,
    );
  }
}

function parseErrorReport(output) {
  const report = {};
  for (const line of output.split(/\r?\n/)) {
    const match = line.match(/^([a-z_]+):\s*(.*)$/);
    if (match) {
      report[match[1]] = match[2];
    }
  }
  return report;
}

function compareErrorReports(label, bashOutput, powerShellOutput, replacements) {
  const bashReport = parseErrorReport(bashOutput);
  const powerShellReport = parseErrorReport(powerShellOutput);
  for (const key of [
    "operation",
    "target_path",
    "selector",
    "current_head",
    "worktree_state",
    "network_state",
    "cause",
    "next_step",
  ]) {
    const bashValue = normalizedValue(bashReport, key, replacements);
    const powerShellValue = normalizedValue(powerShellReport, key, replacements);
    assert(
      bashValue === powerShellValue,
      `${label} error-report mismatch for ${key}: Bash=${bashValue} PowerShell=${powerShellValue}`,
    );
  }
}

function checkoutState(target) {
  const lockPath = `${target}.lock`;
  return {
    head: runGit(target, ["rev-parse", "HEAD"]),
    branch: runGit(target, ["rev-parse", "--abbrev-ref", "HEAD"]),
    shallow: runGit(target, ["rev-parse", "--is-shallow-repository"]),
    origin: runGit(target, ["config", "--get", "remote.origin.url"]),
    status: runGit(target, ["status", "--porcelain"]),
    lock: fs
      .readFileSync(lockPath, "utf8")
      .replace(/^checked_at:.*$/m, "checked_at: <TIME>"),
  };
}

function compareCheckoutStates(label, bashState, powerShellState) {
  for (const key of ["head", "branch", "shallow", "origin", "status", "lock"]) {
    assert(
      bashState[key] === powerShellState[key],
      `${label} state mismatch for ${key}: Bash=${bashState[key]} PowerShell=${powerShellState[key]}`,
    );
  }
}

function runCloneLifecycleParity(bash, powerShell, env) {
  const fixtureRoot = support.createFixtureDirectory("modding-api-surface-parity-");
  try {
    const remote = path.join(fixtureRoot, "modding-api.git");
    const seed = path.join(fixtureRoot, "seed");
    const metadataFile = path.join(fixtureRoot, "latest.json");
    const bashTarget = path.join(fixtureRoot, "bash-reference");
    const powerShellTarget = path.join(fixtureRoot, "powershell-reference");
    runGit(fixtureRoot, ["init", "--bare", remote]);
    runGit(fixtureRoot, ["init", seed]);
    runGit(seed, ["config", "user.email", "test@example.invalid"]);
    runGit(seed, ["config", "user.name", "ModdingAPI parity test"]);
    fs.writeFileSync(path.join(seed, "README.md"), "stable\n", "utf8");
    runGit(seed, ["add", "README.md"]);
    runGit(seed, ["commit", "-m", "initial stable reference"]);
    runGit(seed, ["branch", "-M", "main"]);
    runGit(seed, ["tag", "-a", "v1.0.0", "-m", "stable release"]);
    runGit(seed, ["remote", "add", "origin", remote]);
    runGit(seed, ["push", "--set-upstream", "origin", "main", "--tags"]);
    const releaseCommit = runGit(seed, ["rev-parse", "refs/tags/v1.0.0^{commit}"]);
    fs.writeFileSync(
      metadataFile,
      JSON.stringify({
        tag_name: "v1.0.0",
        draft: false,
        prerelease: false,
        resolved_ref: "v1.0.0",
        resolved_commit: releaseCommit,
      }),
      "utf8",
    );

    const fixtureEnv = Object.assign({}, env, {
      MODDING_API_TEST_MODE: "1",
      MODDING_API_TEST_REPOSITORY: remote,
    });
    const cloneShell = path.join(scriptDirectory, "clone_modding_api.sh");
    const clonePowerShell = path.join(scriptDirectory, "clone_modding_api.ps1");
    const managerShell = path.join(scriptDirectory, "manage_modding_api.sh");
    const managerPowerShell = path.join(scriptDirectory, "manage_modding_api.ps1");
    const cloneBash = support.invokeBash(
      bash,
      cloneShell,
      ["--target-path", bashTarget, "--selector", "latest", "--metadata-file", metadataFile],
      fixtureEnv,
    );
    const clonePowerShellResult = support.invokePowerShell(
      powerShell,
      clonePowerShell,
      ["-TargetPath", powerShellTarget, "-Selector", "latest", "-MetadataFile", metadataFile],
      fixtureEnv,
    );
    assert(cloneBash.status === 0 && clonePowerShellResult.status === 0, "clone surface parity fixture failed");
    compareSurfaceValues(
      "clone",
      cloneBash.output,
      clonePowerShellResult.output,
      [
        "MODDING_API_OPERATION",
        "MODDING_API_REFERENCE_PATH",
        "MODDING_API_SELECTOR",
        "MODDING_API_SELECTOR_KIND",
        "MODDING_API_RESOLVED_REF",
        "MODDING_API_RESOLVED_TAG",
        "MODDING_API_RESOLVED_COMMIT",
        "MODDING_API_SHALLOW",
        "MODDING_API_LOCK_PATH",
      ],
      [
        [bashTarget, "<TARGET>"],
        [powerShellTarget, "<TARGET>"],
        [`${bashTarget}.lock`, "<TARGET>.lock"],
        [`${powerShellTarget}.lock`, "<TARGET>.lock"],
      ],
    );
    compareCheckoutStates(
      "clone",
      checkoutState(bashTarget),
      checkoutState(powerShellTarget),
    );

    const checkBash = support.invokeBash(
      bash,
      managerShell,
      ["--operation", "check", "--target-path", bashTarget, "--selector", "latest", "--metadata-file", metadataFile, "--offline"],
      fixtureEnv,
    );
    const checkPowerShell = support.invokePowerShell(
      powerShell,
      managerPowerShell,
      ["-Operation", "check", "-TargetPath", powerShellTarget, "-Selector", "latest", "-MetadataFile", metadataFile, "-Offline"],
      fixtureEnv,
    );
    assert(checkBash.status === 0 && checkPowerShell.status === 0, "offline check surface parity fixture failed");
    compareSurfaceValues(
      "offline check",
      checkBash.output,
      checkPowerShell.output,
      [
        "MODDING_API_OPERATION",
        "MODDING_API_REFERENCE_PATH",
        "MODDING_API_SELECTOR",
        "MODDING_API_SELECTOR_KIND",
        "MODDING_API_RESOLVED_REF",
        "MODDING_API_RESOLVED_TAG",
        "MODDING_API_RESOLVED_COMMIT",
        "MODDING_API_NETWORK",
        "MODDING_API_DRY_RUN",
        "MODDING_API_LOCK_MATCH",
        "MODDING_API_CHECKOUT_CHANGED",
        "MODDING_API_LOCK_UPDATED",
      ],
      [
        [bashTarget, "<TARGET>"],
        [powerShellTarget, "<TARGET>"],
      ],
    );
    compareCheckoutStates(
      "offline check",
      checkoutState(bashTarget),
      checkoutState(powerShellTarget),
    );

    fs.writeFileSync(path.join(bashTarget, "README.md"), "dirty\n", "utf8");
    fs.writeFileSync(path.join(powerShellTarget, "README.md"), "dirty\n", "utf8");
    const updateBash = support.invokeBash(
      bash,
      managerShell,
      ["--operation", "update", "--target-path", bashTarget, "--selector", "latest", "--metadata-file", metadataFile],
      fixtureEnv,
    );
    const updatePowerShell = support.invokePowerShell(
      powerShell,
      managerPowerShell,
      ["-Operation", "update", "-TargetPath", powerShellTarget, "-Selector", "latest", "-MetadataFile", metadataFile],
      fixtureEnv,
    );
    assert(updateBash.status === 1 && updatePowerShell.status === 1, "dirty update exit-code parity failed");
    compareErrorReports("dirty update", updateBash.output, updatePowerShell.output, [
      [bashTarget, "<TARGET>"],
      [powerShellTarget, "<TARGET>"],
    ]);
    compareCheckoutStates(
      "dirty update",
      checkoutState(bashTarget),
      checkoutState(powerShellTarget),
    );
  } finally {
    fs.rmSync(fixtureRoot, { recursive: true, force: true });
  }
}

function runInstallerSmoke(env) {
  const installer = path.join(repositoryRoot, "bin", "install.js");
  for (const agent of ["trae-cn", "claude-code"]) {
    const result = support.runCommand(
      process.execPath,
      [installer, "--dry-run", "--only", agent],
      { env },
    );
    if (result.status !== 0) {
      fail(`installer dry-run for ${agent} failed:\n${result.output}`);
    }
  }
  const helpResult = support.runCommand(process.execPath, [installer, "--help"], { env });
  if (helpResult.status !== 0) {
    fail(`installer help failed:\n${helpResult.output}`);
  }
}

function gitMarkdownFiles(args) {
  const result = support.runCommand("git", args);
  assert(result.status === 0, `Git Markdown file discovery failed:\n${result.output}`);
  return result.output
    .split("\0")
    .filter((relativePath) => relativePath.length > 0)
    .filter((relativePath) => relativePath.toLowerCase().endsWith(".md"))
    .map((relativePath) => path.resolve(repositoryRoot, relativePath))
    .filter((file) => fs.existsSync(file) && fs.statSync(file).isFile());
}

function trackedMarkdownFiles() {
  return gitMarkdownFiles([
    "ls-files",
    "--cached",
    "--others",
    "--exclude-standard",
    "-z",
    "--",
    "*.md",
  ]);
}

function ignoredMarkdownFiles() {
  return gitMarkdownFiles([
    "ls-files",
    "--others",
    "--ignored",
    "--exclude-standard",
    "-z",
    "--",
    "*.md",
  ]);
}

function displayPath(file) {
  return path.relative(repositoryRoot, file).replace(/\\/g, "/");
}

function markdownTarget(rawTarget) {
  let target = rawTarget.trim();
  if (target.startsWith("<") && target.includes(">")) {
    target = target.slice(1, target.indexOf(">"));
  } else {
    target = target.split(/\s+/)[0];
  }
  return target.split("#", 1)[0];
}

function findMissingMarkdownLinks(files) {
  const missing = [];
  for (const file of files) {
    const contents = fs.readFileSync(file, "utf8");
    const linkPattern = /!?\[[^\]]*\]\(([^)\r\n]+)\)/g;
    let match;
    while ((match = linkPattern.exec(contents)) !== null) {
      const target = markdownTarget(match[1]);
      if (
        !target ||
        target.startsWith("#") ||
        /^[A-Za-z][A-Za-z0-9+.-]*:/.test(target) ||
        target.startsWith("//")
      ) {
        continue;
      }
      const resolved = path.resolve(path.dirname(file), decodeURIComponent(target));
      if (!fs.existsSync(resolved)) {
        missing.push(`${displayPath(file)} -> ${target}`);
      }
    }
  }
  return missing;
}

function checkMarkdownLinks() {
  const repositoryFiles = trackedMarkdownFiles();
  const ignoredFiles = ignoredMarkdownFiles();
  const missing = findMissingMarkdownLinks(repositoryFiles);
  const ignoredMissing = findMissingMarkdownLinks(ignoredFiles);

  if (ignoredFiles.length > 0) {
    process.stdout.write(
      `[INFO] ignored Markdown files excluded from repository link validation (${ignoredFiles.length}):\n` +
        `${ignoredFiles.map(displayPath).join("\n")}\n`,
    );
  }
  if (ignoredMissing.length > 0) {
    process.stdout.write(
      `[WARN] ignored local Markdown link findings (not release failures):\n` +
        `${ignoredMissing.join("\n")}\n`,
    );
  }
  assert(missing.length === 0, `missing repository Markdown links:\n${missing.join("\n")}`);
}

function checkGitDiff() {
  const result = support.runCommand("git", ["diff", "--check"]);
  assert(result.status === 0, `git diff --check failed:\n${result.output}`);
}

function checkCleanWorktree() {
  const result = support.runCommand("git", ["status", "--porcelain"]);
  assert(result.status === 0, `git status failed:\n${result.output}`);
  assert(!result.output.trim(), `worktree is not clean:\n${result.output}`);
}

function run() {
  const options = parseArguments(process.argv.slice(2));
  if (options === null) {
    return;
  }
  const bash = support.findBash();
  const powerShell = support.findPowerShell();
  const env = Object.assign({}, process.env, {
    MODDING_API_POWERSHELL: powerShell,
  });
  runResolverParity(bash, powerShell, env);
  runSuite("clone/lifecycle surface parity", () => {
    runCloneLifecycleParity(bash, powerShell, env);
    return { status: 0, output: "" };
  });

  const suites = [
    ["resolver Bash", "bash", "test_resolve_modding_api.sh"],
    ["resolver PowerShell", "powershell", "test_resolve_modding_api.ps1"],
    ["clone Bash", "bash", "test_clone_modding_api.sh"],
    ["clone PowerShell", "powershell", "test_clone_modding_api.ps1"],
    ["lifecycle Bash", "bash", "test_manage_modding_api.sh"],
    ["lifecycle PowerShell", "powershell", "test_manage_modding_api.ps1"],
    ["documentation Bash", "bash", "test_referencing_modding_api.sh"],
    ["documentation PowerShell", "powershell", "test_referencing_modding_api.ps1"],
  ];
  for (const [label, surface, fileName] of suites) {
    const script = path.join(scriptDirectory, fileName);
    runSuite(label, () =>
      surface === "bash"
        ? support.invokeBash(bash, script, [], env)
        : support.invokePowerShell(powerShell, script, [], env),
    );
  }

  runSuite("installer dry-run smoke", () => {
    runInstallerSmoke(env);
    return { status: 0, output: "" };
  });
  runSuite("Markdown link check", () => {
    checkMarkdownLinks();
    return { status: 0, output: "" };
  });
  runSuite("git diff --check", () => {
    checkGitDiff();
    return { status: 0, output: "" };
  });
  if (options.requireClean) {
    runSuite("clean worktree check", () => {
      checkCleanWorktree();
      return { status: 0, output: "" };
    });
  }
  process.stdout.write("[OK] ModdingAPI cross-platform acceptance gate\n");
}

try {
  run();
} catch (error) {
  process.stderr.write(`[FAIL] ${error.message}\n`);
  process.exitCode = 1;
}

File: skills/blasphemous-modding-helper/scripts/test_modding_api_acceptance.py
#!/usr/bin/env python3
"""Run the Python-only deterministic acceptance surface for the Skill."""

from __future__ import annotations

import argparse
import os
import platform
import re
import shutil
import subprocess
import sys
from pathlib import Path
from typing import Callable, Iterable, List, Optional, Sequence
from urllib.parse import unquote


SCRIPT_ROOT = Path(__file__).resolve().parent
SKILL_ROOT = SCRIPT_ROOT.parent
REPOSITORY_ROOT = SKILL_ROOT.parents[1]
ROOT_RUNNER = REPOSITORY_ROOT / "tests" / "run_blasphemous_modding_test.py"
INSTALLER = REPOSITORY_ROOT / "bin" / "install.js"
INSTALLER_TEST = REPOSITORY_ROOT / "tests" / "test_installer.js"


class AcceptanceError(RuntimeError):
    """A user-facing acceptance-gate failure."""


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        description=(
            "Run the Python-only deterministic acceptance surface for the "
            "Blasphemous ModdingAPI Skill."
        ),
        epilog=(
            "Migrated Skill behavior runs through Python entry points. "
            "The Node.js installer is tested separately; live GitHub, Steam, "
            "game, and Manual verification remain opt-in."
        ),
    )
    parser.add_argument(
        "--python",
        dest="python_executable",
        default=sys.executable,
        help="Python 3 executable for the deterministic test surface (default: current interpreter).",
    )
    parser.add_argument(
        "--node",
        dest="node_executable",
        default=os.environ.get("MODDING_API_NODE") or shutil.which("node") or "node",
        help="Node.js executable for the separate installer boundary tests.",
    )
    parser.add_argument(
        "--require-clean",
        action="store_true",
        help="Fail after all checks unless the repository worktree is clean.",
    )
    return parser


def run_command(
    label: str,
    command: Sequence[str],
    *,
    cwd: Path = REPOSITORY_ROOT,
) -> subprocess.CompletedProcess[str]:
    print(f"[RUN] {label}")
    try:
        result = subprocess.run(
            list(command),
            cwd=str(cwd),
            env=os.environ.copy(),
            capture_output=True,
            text=True,
            encoding="utf-8",
            errors="replace",
            shell=False,
            check=False,
        )
    except OSError as error:
        raise AcceptanceError(f"{label} could not start: {error}") from error
    if result.stdout:
        print(result.stdout, end="")
    if result.stderr:
        print(result.stderr, end="", file=sys.stderr)
    if result.returncode != 0:
        raise AcceptanceError(
            f"{label} failed with exit code {result.returncode}."
        )
    print(f"[OK] {label}")
    return result


def run_check(label: str, check: Callable[[], None]) -> None:
    print(f"[RUN] {label}")
    try:
        check()
    except (AcceptanceError, OSError, UnicodeError) as error:
        raise AcceptanceError(f"{label} failed: {error}") from error
    print(f"[OK] {label}")


def run_python_surface(python_executable: str) -> None:
    run_command(
        "Python deterministic contract surface",
        (
            python_executable,
            str(ROOT_RUNNER),
            "--python",
            python_executable,
        ),
    )


def run_installer_surface(node_executable: str) -> None:
    run_command(
        "Node installer regression test",
        (node_executable, str(INSTALLER_TEST)),
    )
    for agent in ("trae-cn", "claude-code"):
        run_command(
            f"Node installer dry-run ({agent})",
            (node_executable, str(INSTALLER), "--dry-run", "--only", agent),
        )
    run_command("Node installer help", (node_executable, str(INSTALLER), "--help"))


def _git_markdown_files(arguments: Sequence[str]) -> List[Path]:
    try:
        result = subprocess.run(
            ["git", *arguments],
            cwd=str(REPOSITORY_ROOT),
            env=os.environ.copy(),
            capture_output=True,
            text=True,
            encoding="utf-8",
            errors="replace",
            shell=False,
            check=False,
        )
    except OSError as error:
        raise AcceptanceError(f"Git Markdown file discovery could not start: {error}") from error
    if result.returncode != 0:
        raise AcceptanceError(
            "Git Markdown file discovery failed:\n"
            f"{result.stdout}\n{result.stderr}"
        )
    paths = []
    for raw_path in result.stdout.split("\0"):
        if not raw_path or not raw_path.lower().endswith(".md"):
            continue
        path = (REPOSITORY_ROOT / raw_path).resolve()
        if path.is_file():
            paths.append(path)
    return paths


def tracked_markdown_files() -> List[Path]:
    return _git_markdown_files(
        (
            "ls-files",
            "--cached",
            "--others",
            "--exclude-standard",
            "-z",
            "--",
            "*.md",
        )
    )


def ignored_markdown_files() -> List[Path]:
    return _git_markdown_files(
        (
            "ls-files",
            "--others",
            "--ignored",
            "--exclude-standard",
            "-z",
            "--",
            "*.md",
        )
    )


def display_path(path: Path) -> str:
    return path.relative_to(REPOSITORY_ROOT).as_posix()


def markdown_target(raw_target: str) -> str:
    target = raw_target.strip()
    if target.startswith("<") and ">" in target:
        target = target[1 : target.index(">")]
    else:
        target = target.split()[0] if target.split() else ""
    return target.split("#", 1)[0]


def find_missing_markdown_links(files: Iterable[Path]) -> List[str]:
    missing: List[str] = []
    link_pattern = re.compile(r"!?\[[^\]]*\]\(([^)\r\n]+)\)")
    for path in files:
        contents = path.read_text(encoding="utf-8")
        for match in link_pattern.finditer(contents):
            target = markdown_target(match.group(1))
            if (
                not target
                or target.startswith("#")
                or re.match(r"^[A-Za-z][A-Za-z0-9+.-]*:", target)
                or target.startswith("//")
            ):
                continue
            resolved = (path.parent / unquote(target)).resolve()
            if not resolved.exists():
                missing.append(f"{display_path(path)} -> {target}")
    return missing


def check_markdown_links() -> None:
    repository_files = tracked_markdown_files()
    ignored_files = ignored_markdown_files()
    missing = find_missing_markdown_links(repository_files)
    ignored_missing = find_missing_markdown_links(ignored_files)
    if ignored_files:
        print(
            "[INFO] ignored Markdown files excluded from repository link "
            f"validation ({len(ignored_files)}):"
        )
        for path in ignored_files:
            print(display_path(path))
    if ignored_missing:
        print("[WARN] ignored local Markdown link findings (not release failures):")
        for item in ignored_missing:
            print(item)
    if missing:
        raise AcceptanceError(
            "missing repository Markdown links:\n" + "\n".join(missing)
        )


def check_git_diff() -> None:
    run_command("git diff --check", ("git", "diff", "--check"))


def check_clean_worktree() -> None:
    result = run_command("clean worktree check", ("git", "status", "--porcelain"))
    if result.stdout.strip():
        raise AcceptanceError(f"worktree is not clean:\n{result.stdout}")


def main(argv: Optional[Sequence[str]] = None) -> int:
    args = build_parser().parse_args(argv)
    print("Acceptance runner: native Python")
    print(f"Platform: {platform.system()}")
    print(f"Python: {args.python_executable}")
    print(f"Node installer boundary: {args.node_executable}")
    try:
        run_python_surface(args.python_executable)
        run_installer_surface(args.node_executable)
        run_check("Markdown link check", check_markdown_links)
        run_check("git diff --check", check_git_diff)
        if args.require_clean:
            run_check("clean worktree check", check_clean_worktree)
    except AcceptanceError as error:
        print(f"[FAIL] {error}", file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
